kotlin/samples/hello-terminal/terminal-workflow/src/main/java/com/squareup/sample/helloterminal/terminalworkflow/KeyStroke.kt

```diff
@@ -55,19 +60,72 @@ data class KeyStroke(
   }
 }
 
+/**
+ * Returns a new [Flow] of keystrokes from this input provider. The flow will _not_ be multicasted,
+ * so you'll want to do that yourself or ensure that there is only ever one collector at a time.
+ * You will also probably want to specify an IO dispatcher to run it on, since it does blocking
+ * IO.
+ *
+ * This is a function instead of a property because calling it multiple times will return
+ * independent flows.
+ */
 @Suppress("BlockingMethodInNonBlockingContext")
 @UseExperimental(ExperimentalCoroutinesApi::class)
-internal fun InputProvider.listenForKeyStrokesOn(
-  scope: CoroutineScope
-): BroadcastChannel<KeyStroke> = scope.broadcast {
-  while (isActive) {
-    val keyStroke = readInput()
-    if (keyStroke.keyType === EOF) {
-      // EOF indicates the terminal input was closed, and we won't receive any more input, so
-      // close the channel instead of sending the raw event down.
-      return@broadcast
+internal fun InputProvider.keyStrokes(): Flow<KeyStroke> = flow {
+  coroutineScope {
+    while (isActive) {
+      // Since readInput doesn't take a timeout, we force it by interrupting the blocked thread
+      // when we get cancelled.
+      val keyStroke = interruptThreadOnCancel { readInput() }
+      if (keyStroke.keyType === EOF) {
+        // EOF indicates the terminal input was closed, and we won't receive any more input, so
+        // close the channel instead of sending the raw event down.
+        break
+      }
+      emit(keyStroke.toKeyStroke())
+    }
+  }
+}
+
+/**
+ * Sends and interrupt to the calling thread when the [CoroutineScope] is cancelled. Intended to
+ * make cancellation work correctly when [block] performs blocking IO – the blocked thread will not
+ * get the cancellation signal, so we interrupt it to cancel the IO, then catch the interrupt
+ * exception and throw the cancellation exception instead.
+ */
+@ExperimentalCoroutinesApi
+private fun <T> CoroutineScope.interruptThreadOnCancel(block: () -> T): T {
```

Comment:
This was fun to solve, although I'm not sure how I feel about this solution. The nice thing is that it should work with just about any JVM IO as long as it supports interrupts (I think all of it does?). The gross thing is that it launches a new coroutine for every iteration, but we could get around that using an internal API (see the comments).

Reply:
They've been bragging so much about how cheap coroutines are, I feel good putting it to the test. Especially in a sample.